go/adapter/datadog/adapter.go

```diff
@@ -43,76 +39,54 @@ func NewDatadogAdapter(config *DatadogConfig) (DatadogAdapter, error) {
 
 	return DatadogAdapter{
 		AdapterBase: observe.NewAdapterBase(),
-		TraceId:     uint64(observe.NewTraceId()),
 		Config:      config,
 	}, nil
 }
 
-func (d *DatadogAdapter) Event(e observe.Event) {
-	switch event := e.(type) {
-	case observe.CallEvent:
-		spans := d.makeCallSpans(event, nil)
-		if len(spans) > 0 {
-			d.Spans = append(d.Spans, spans...)
-		}
-
-	case observe.MemoryGrowEvent:
-		if len(d.Spans) > 0 {
-			d.Spans[len(d.Spans)-1].AddAllocation(event.MemoryGrowAmount())
-		}
-	case observe.CustomEvent:
-		if value, ok := event.Metadata["trace_id"]; ok {
-			traceId, err := strconv.ParseUint(value.(string), 10, 64)
-			if err != nil {
-				log.Println("failed to parse traceId from event metadata:", err)
-				return
-			}
-
-			d.TraceId = traceId
-		}
-	}
+func (d *DatadogAdapter) Start() {
+	d.AdapterBase.Start(d)
 }
 
-func (d *DatadogAdapter) Wait(collector *observe.Collector, timeout time.Duration) {
-	d.AdapterBase.Wait(collector, timeout, nil)
+func (d *DatadogAdapter) Stop() {
+	d.AdapterBase.Stop()
 }
 
-func (d *DatadogAdapter) Start(collector *observe.Collector, wasm []byte) error {
-	if err := d.AdapterBase.Start(collector, wasm); err != nil {
-		return err
+func (d *DatadogAdapter) HandleTraceEvent(te observe.TraceEvent) {
+	if te.TelemetryId == nil {
+		log.Println("Datadog adapter needs a trace id")
+		return
 	}
 
-	stop := d.StopChan(collector)
-
-	go func() {
-		for {
-			select {
-			case event := <-collector.Events:
-				d.Event(event)
-			case <-stop:
-				return
+	var allSpans []datadog_formatter.Span
+	for _, e := range te.Events {
+		switch event := e.(type) {
+		case observe.CallEvent:
+			spans := d.makeCallSpans(event, nil, *te.TelemetryId)
+			if len(spans) > 0 {
+				allSpans = append(allSpans, spans...)
 			}
+		case observe.MemoryGrowEvent:
+			if len(allSpans) > 0 {
+				allSpans[len(allSpans)-1].AddAllocation(event.MemoryGrowAmount())
+			}
+		case observe.CustomEvent:
+			log.Println("Datadog adapter does not respect custom events")
```

Comment:
Idk if we need to log this -- could result in a ton of noise, but will leave it up to you.

Reply:
going to think about it. I don't this we even need this event type anymore.

Reply:
Yea this was possibly a placeholder to support things like setting metadata.. but the design may have changed enough to handle that another way?

Reply:
Yes, we don't need to pass that stuff through events in Go because we can share memory. In my next ticket where I do the metadata stuff in JS and Go, i'll demonstrate. Or refactor at that point.